drivers/display/display_st7735r.c

```diff
@@ -0,0 +1,461 @@
+/*
+ * Copyright (c) 2020 Kim Bøndergaard <kim@fam-boendergaard.dk>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT sitronix_st7735r
+
+#include "display_st7735r.h"
+
+#include <device.h>
+#include <drivers/spi.h>
+#include <drivers/gpio.h>
+#include <sys/byteorder.h>
+#include <drivers/display.h>
+
+#define LOG_LEVEL CONFIG_DISPLAY_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(display_st7735r);
+
+#define ST7735R_CS_PIN		DT_INST_SPI_DEV_CS_GPIOS_PIN(0)
+#define ST7735R_CMD_DATA_PIN	DT_INST_GPIO_PIN(0, cmd_data_gpios)
+#define ST7735R_CMD_DATA_FLAGS	DT_INST_GPIO_FLAGS(0, cmd_data_gpios)
+#define ST7735R_RESET_PIN	DT_INST_GPIO_PIN(0, reset_gpios)
+#define ST7735R_RESET_FLAGS	DT_INST_GPIO_FLAGS(0, reset_gpios)
+
+static uint8_t st7735r_pwctrl1_param[] = DT_INST_PROP(0, pwctrl1_param);
+static uint8_t st7735r_pvgam_param[] = DT_INST_PROP(0, pvgam_param);
+static uint8_t st7735r_nvgam_param[] = DT_INST_PROP(0, nvgam_param);
+
+struct st7735r_data {
+	const struct device *spi_dev;
+	struct spi_config spi_config;
+#if DT_INST_SPI_DEV_HAS_CS_GPIOS(0)
+	struct spi_cs_control cs_ctrl;
+#endif
+
+#if DT_INST_NODE_HAS_PROP(0, reset_gpios)
+	const struct device *reset_gpio;
+#endif
+	const struct device *cmd_data_gpio;
+
+	uint16_t height;
+	uint16_t width;
+	uint16_t x_offset;
+	uint16_t y_offset;
+#ifdef CONFIG_DEVICE_POWER_MANAGEMENT
+	uint32_t pm_state;
+#endif
+};
+
+#define ST7735R_PIXEL_SIZE 2u
+
+static void st7735r_set_lcd_margins(struct st7735r_data *data,
+			     uint16_t x_offset, uint16_t y_offset)
+{
+	data->x_offset = x_offset;
+	data->y_offset = y_offset;
+}
+
+static void st7735r_set_cmd(struct st7735r_data *data, int is_cmd)
+{
+	gpio_pin_set(data->cmd_data_gpio, ST7735R_CMD_DATA_PIN, is_cmd);
+}
+
+static void st7735r_transmit(struct st7735r_data *data, uint8_t cmd,
+		uint8_t *tx_data, size_t tx_count)
+{
+	struct spi_buf tx_buf = { .buf = &cmd, .len = 1 };
+	struct spi_buf_set tx_bufs = { .buffers = &tx_buf, .count = 1 };
+
+	st7735r_set_cmd(data, 1);
+	spi_write(data->spi_dev, &data->spi_config, &tx_bufs);
+
+	if (tx_data != NULL) {
+		tx_buf.buf = tx_data;
+		tx_buf.len = tx_count;
+		st7735r_set_cmd(data, 0);
+		spi_write(data->spi_dev, &data->spi_config, &tx_bufs);
+	}
+}
+
+static void st7735r_exit_sleep(struct st7735r_data *data)
+{
+	st7735r_transmit(data, ST7735R_CMD_SLEEP_OUT, NULL, 0);
+	k_sleep(K_MSEC(120));
+}
+
+static void st7735r_reset_display(struct st7735r_data *data)
+{
+	LOG_DBG("Resetting display");
+#if DT_INST_NODE_HAS_PROP(0, reset_gpios)
+	k_sleep(K_MSEC(1));
+	gpio_pin_set(data->reset_gpio, ST7735R_RESET_PIN, 1);
+	k_sleep(K_MSEC(6));
+	gpio_pin_set(data->reset_gpio, ST7735R_RESET_PIN, 0);
+	k_sleep(K_MSEC(20));
```

Comment:
Magic constants should be defined, e.g. `ST7735R_WAKEUP_TIME`

Reply:
:+1: